js/src/components/tree-select-control/index.js

```diff
@@ -1,56 +1,215 @@
+				repository[ child.id ] = { ...child };
+
+				if ( child.children ) {
+					loadOption( child );
+				}
+			} );
+		}
+
+		optionsRef.forEach( ( option ) => loadOption( option ) );
+
+		return repository;
+	}, [ optionsRef ] );
+
+	/**
+	 * Get formatted Tags from the selected values.
+	 *
+	 * @return {{name: {string}, id: {string}}[]} An array of Tags
+	 */
+	const getTags = () => {
+		if ( ! options.length ) {
+			return [];
+		}
+
+		return value.map( ( key ) => {
+			const option = optionsRepository[ key ];
+			return { id: key, name: option.name };
+		} );
+	};
+
+	/**
+	 * Handles a change on the Tree List of options. Could be a click on a parent option
+	 * or a child option
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleListChange = ( checked, option ) => {
+		if ( option.children?.length ) {
+			handleParentChange( checked, option );
+		} else {
+			handleSingleChange( checked, option );
+		}
+	};
+
+	/**
+	 * Handles a change of a child element.
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleSingleChange = ( checked, option ) => {
+		const idPosition = value.indexOf( option.id );
+		const newValue = [ ...value ];
+
+		if ( ! checked ) {
+			newValue.splice( idPosition, 1 );
+		} else {
+			newValue.push( option.id );
+		}
+
+		onChange( newValue );
+	};
+
+	/**
+	 * Handles a change of a Parent element.
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleParentChange = ( checked, option ) => {
+		const newValue = [ ...value ];
+
+		function loadChildren( parent ) {
+			if ( ! parent.children ) {
+				return;
+			}
+
+			parent.children.forEach( ( child ) => {
+				if ( child.children?.length ) {
+					loadChildren( child );
+					return;
+				}
+
+				const childIdPosition = newValue.indexOf( child.id );
+
+				if ( ! checked && childIdPosition >= 0 ) {
+					newValue.splice( childIdPosition, 1 );
+				}
+
+				if ( checked && childIdPosition < 0 ) {
+					newValue.push( child.id );
+				}
+			} );
+		}
+
+		loadChildren( option );
+		onChange( newValue );
+	};
+
+	/**
+	 * Handles a change of a Tag element. We map them to Value format.
+	 *
+	 * @param {Array} tags List of current tags
+	 */
+	const handleTagsChange = ( tags ) => {
+		onChange( [ ...tags.map( ( el ) => el.id ) ] );
+	};
 
 	return (
 		<div
+			{ ...focusOutside }
 			className={ classnames(
 				'woocommerce-tree-select-control',
 				className
 			) }
 		>
+			{ !! label && (
+				<label
+					htmlFor={ `woocommerce-tree-select-control-${ id }__control-input` }
+					className="woocommerce-tree-select-control__label"
+				>
+					{ label }
+				</label>
+			) }
+
 			<Control
-				disabled={ disabled }
-				hasTags
+				disabled={ disabled || ! options.length }
```

Comment:
❔ Do we need to disable it when there are no options?

Reply:
I don't find any advantage in allowing the user to type or click into a search box without options. Besides the fact that it will rely on more code for handling exceptions caused by potential iterations to options array for the query search and other features based on options. 

Not sure if it will be a case of we need the selector with "no options" but definitely disabling the control on that case will help to minimize the edge cases created by that.

Reply:
Okay, I guess you are thinking from the defensive code safety point of view. However, from the component consumer point of view, I would feel like "I didn't specify `disabled`, so why is this TreeSelect control rendered as disabled?", "I want it to be disabled only when I explicitly put in the code `disabled`", and it _might_ end up being a bug like "TreeSelect is disabled when there are no options" or a feature request. 😅 

For reference, here's the Ant Design TreeSelect component with `treeData={[]}`:

<img width="776" alt="image" src="https://user-images.githubusercontent.com/417342/155993517-e41d066e-f5cc-4a38-95c0-65280f18c342.png">

Even though it is pretty much useless without the options, it would be the consumer's responsibility to figure out what happen, and if they want to disable it, they can code it in like `<TreeSelect disabled={ treeData.length === 0 }>`. The consumer has the choice. With the code we have now, the consumer doesn't have the choice but to accept our way.

Reply:
Added here https://github.com/woocommerce/google-listings-and-ads/pull/1243/commits/f00ff223bb1ece72bf59d8e64172fba90e083ac3